### PR TITLE
feat:  resources dereference [DEV-1584]

### PR DESCRIPTION
**Pull request by Toktar**

### CODE REVIEW

File: services/resource_dereference_service.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import (
+	// jsonpb Marshaller is deprecated, but is needed because there's only one way to proto
+	// marshal in combination with our proto generator version
+	"encoding/json"
+
+	"github.com/cheqd/did-resolver/types"
+	"github.com/cheqd/did-resolver/utils"
+)
+
+type ResourceDereferenceService struct {
```
Comment (askolesov) [inline services/resource_dereference_service.go:12]: As a general idea. What do you think about addin interfaces to all services? Just to make clearer what the method list is.

Comment (askolesov) [reply services/resource_dereference_service.go:12]: The same way it's done for LedgerService

File: services/resource_dereference_service.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	ledgerService LedgerServiceI
+	didDocService DIDDocService
```
Comment (askolesov) [inline services/resource_dereference_service.go:14]: May be not depend on didDoc service here? It doesn't seem like we need anything DidDoc related here. The service is used for content stream marshalling only which is pretty generic. What do you think about extracting this method into let's say `resolution_service` or helpers?

File: services/resource_dereference_service.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+func NewResourceDereferenceService(ledgerService LedgerServiceI, didDocService DIDDocService) ResourceDereferenceService {
+	return ResourceDereferenceService{
+		ledgerService: ledgerService,
+		didDocService: didDocService,
+	}
+}
+
+func (rds ResourceDereferenceService) DereferenceResource(path string, did string, didUrl string, dereferenceOptions types.DereferencingOption) (types.DidDereferencing, int) {
```
Comment (askolesov) [inline services/resource_dereference_service.go:24]: Why do we need path and did here? Can't we extract them from didUrl?

File: services/resource_dereference_service.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	if utils.IsResourceHeaderPath(path) {
+		jsonCotentStream, dereferenceMetadata = rds.dereferenceHeader(path, did, dereferenceOptions)
+	} else if utils.IsCollectionResourcesPath(path) || utils.IsCollectionResourcesPathRedirect(path) {
+		jsonCotentStream, dereferenceMetadata = rds.dereferenceCollectionResources(path, did, dereferenceOptions)
+	} else {
+		dereferenceMetadata.ResolutionError = types.RepresentationNotSupportedError
+	}
+
+	var statusCode int
+	if utils.IsCollectionResourcesPathRedirect(path) {
```
Comment (askolesov) [inline services/resource_dereference_service.go:37]: Yeah, this nieeds to be improved, just a riminder fo me

File: services/resource_dereference_service.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	resourceId := utils.GetResourceId(path)
+
+	resource, dereferencingError := rds.ledgerService.QueryResource(did, resourceId)
+
+	if dereferencingError != "" {
+		dereferenceMetadata.ResolutionError = dereferencingError
+		return "", dereferenceMetadata
+	}
+	var err error
+	cotentStream, err := rds.didDocService.MarshallContentStream(resource.Header, dereferenceOptions.Accept)
```
Comment (askolesov) [inline services/resource_dereference_service.go:58]: I think we should either split MarshallContentStream into two methods: for didDocs and for resources, or extract it. Not it appears that didDoc service is aware of resources which is conceptually can be avoided.

Not very important, will leave it up to you.

File: services/resource_dereference_service.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		dereferenceMetadata.ResolutionError = types.InternalError
+	}
+	return cotentStream, dereferenceMetadata
+}
+
+func (rds ResourceDereferenceService) dereferenceCollectionResources(path string, did string, dereferenceOptions types.DereferencingOption) (string, types.DereferencingMetadata) {
```
Comment (askolesov) [inline services/resource_dereference_service.go:65]: Looks like path can be safly removed.

File: services/resource_dereference_service.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	}
+	cotentStream, err := json.Marshal(jsonResources)
+	if err != nil {
+		dereferenceMetadata.ResolutionError = types.InternalError
+		return "", dereferenceMetadata
+	}
+	return string(cotentStream), dereferenceMetadata
+}
+
+func (rds ResourceDereferenceService) DereferenceResourceData(path string, did string, didUrl string, dereferenceOptions types.DereferencingOption) ([]byte, types.DereferencingMetadata) {
```
Comment (askolesov) [inline services/resource_dereference_service.go:89]: The same thing with path and did.

**Comment by askolesov** [comment]
Great work, keep going! Left some comments from the top of my head.

File: services/resource_dereference_service.go (hunk 7 above)
Comment (Toktar) [reply services/resource_dereference_service.go]: Done

File: services/resource_dereference_service.go (hunk 6 above)
Comment (Toktar) [reply services/resource_dereference_service.go]: Done

File: services/resource_dereference_service.go (hunk 4 above)
Comment (Toktar) [reply services/resource_dereference_service.go]: Done

File: services/resource_dereference_service.go (hunk 3 above)
Comment (Toktar) [reply services/resource_dereference_service.go]: Done

**Review by ankurdotb — CHANGES_REQUESTED** [review]
Let's discuss once you've finished the basic fixes but as discussed on Slack. Change your routing, path, and query handling to Echo functions instead. To me this will likely be some of the code below but will likely need significant refactors.

**Review by ankurdotb — APPROVED** [review]

**Pull request merged**